### PR TITLE
SDAP-52 Improve ranking speed

### PR DESCRIPTION
**Pull request by Yongyao**
The time complexity of the current machine learning algorithm O(n2) because the order of some items are not transitive and using collection.sort directly would cause an "comparison method violates it general contract" error. This PR improved it to O(N*logN).

### CODE REVIEW

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

**Comment by lewismc** [comment]
@Yongyao can you please make sure to link the issue title to the JIRA issue title... from JIRA
```SDAP-52 Improve ranking speed```
It makes it easier for us to track the issues in our issue tracker.

File: core/src/main/java/org/apache/sdap/mudrod/ssearch/Ranker.java [hunk 1]
```diff
@@ -147,42 +137,32 @@ private double getNDForm(double d) {
...
-  public int comp(SResult o1, SResult o2) {
-    List<Double> instList = new ArrayList<>();
-    for (int i = 0; i < SResult.rlist.length; i++) {
-      double o2Score = SResult.get(o2, SResult.rlist[i]);
-      double o1Score = SResult.get(o1, SResult.rlist[i]);
-      instList.add(o2Score - o1Score);
-    }
+  public class ResultComparator implements Comparator<SResult> {
+    @Override
+    public int compare(SResult o1, SResult o2) {
```
Comment (lewismc) [inline core/src/main/java/org/apache/sdap/mudrod/ssearch/Ranker.java]: Please document ```SResult o1``` and ```SResult o2``` parameters.

File: core/src/main/java/org/apache/sdap/mudrod/ssearch/Ranker.java [hunk 2]
```diff
@@ -147,42 +137,32 @@ private double getNDForm(double d) {
...
+      List<Double> instList = new ArrayList<>();
+      for (int i = 0; i < SResult.rlist.length; i++) {
```
Comment (lewismc) [inline core/src/main/java/org/apache/sdap/mudrod/ssearch/Ranker.java]: Please use the enhanced for loop notation
```
for (SResult result: SResult.rlist.length)
```
This prevents the additional rlist([i]) accessors which is messy.

File: core/src/main/java/org/apache/sdap/mudrod/ssearch/Ranker.java [hunk 3]
```diff
@@ -147,42 +137,32 @@ private double getNDForm(double d) {
...
-    double prediction = le.classify(insPoint);
-    if (equalComp(prediction, 1)) { //different from weka where the return value is 1 or 2
-      return 0;
-    } else {
-      return 1;
+      double[] ins = instList.stream().mapToDouble(i -> i).toArray();
+      LabeledPoint insPoint = new LabeledPoint(99.0, Vectors.dense(ins));
+      int prediction = (int)le.classify(insPoint);
+      
+      if (prediction==1) { //different from weka where the return value is 1 or 2
```
Comment (lewismc) [inline core/src/main/java/org/apache/sdap/mudrod/ssearch/Ranker.java]: The verified value should ALWAYS be on the left hand side. 
```if(1==prediction)```

**Comment by Yongyao** [comment]
Changes made. @lewismc

File: core/src/main/java/org/apache/sdap/mudrod/ssearch/Ranker.java [hunk 4]
```diff
@@ -147,42 +137,32 @@ private double getNDForm(double d) {
...
+        return 1;
```
Comment (lewismc) [inline core/src/main/java/org/apache/sdap/mudrod/ssearch/Ranker.java]: Would you not just return the variable ```prediction```?

**Review by lewismc — CHANGES_REQUESTED** [review]
Hi @Yongyao this is looking not bad. Please consider my comments. Additionally, do you ave any unit tests to verify the performance improvement? It would be preferred if we submit Unit tests for every patch from now on.

**Review by lewismc — APPROVED** [review]

**Pull request merged**